test/test_license_utils.py

```diff
@@ -0,0 +1,691 @@
+
+    x = LicenseAcceptabilityCheckFailure(unacceptable_licenses=[])
+
+    assert str(x) == LicenseAcceptabilityCheckFailure.DEFAULT_MESSAGE
+
+    alternate_error_message = "Hey! " + LicenseAcceptabilityCheckFailure.DEFAULT_MESSAGE  # different from default
+
+    x = LicenseAcceptabilityCheckFailure(message=alternate_error_message)
+
+    assert str(x) == alternate_error_message
+    assert str(x) != LicenseAcceptabilityCheckFailure.DEFAULT_MESSAGE
+
+    x = LicenseAcceptabilityCheckFailure(unacceptable_licenses=['license1'])
+
+    assert str(x) == "There is 1 unacceptable license: license1"
+
+    x = LicenseAcceptabilityCheckFailure(unacceptable_licenses=['license1'])
+
+    assert str(x) == "There is 1 unacceptable license: license1"
+
+    x = LicenseAcceptabilityCheckFailure(unacceptable_licenses=['license1', 'license2'])
+
+    assert str(x) == "There are 2 unacceptable licenses: license1, license2"
+
+    x = LicenseAcceptabilityCheckFailure(
+        unacceptable_licenses={
+            'license1': ['library1'],
+            'license2': ['library2', 'library3']
+        }
+    )
+
+    assert str(x) == "There are 2 unacceptable licenses: license1, license2"
+
+
+def test_license_status():  # index of status keywords
+
+    for attr in dir(LicenseStatus):
+        if attr.isupper() and not attr.startswith('_'):
+            assert getattr(LicenseStatus, attr) == attr
+
+
+def test_license_analysis():  # represents result of an analysis
+
+    analysis = LicenseAnalysis()
+
+    assert analysis.frameworks == []
+
+    assert analysis.dependency_details == []
+
+    assert analysis.unacceptable == {}
+    assert isinstance(analysis.unacceptable, defaultdict)
+    analysis.unacceptable['some_license'].append('some_library')
+    assert analysis.unacceptable == {'some_license': ['some_library']}
+
+    assert analysis.unexpected_missing == []
+
+    assert analysis.no_longer_missing == []
+
+    assert analysis.miscellaneous == []
+
+
+def test_license_framework():  # class
+
+    assert LicenseFramework.NAME is None
+
+    with pytest.raises(NotImplementedError):
+        LicenseFramework.get_dependencies()
+
+
+def test_license_framework_registry_register():  # decorator
+
+    with pytest.raises(ValueError):
+        @LicenseFrameworkRegistry.register(name='bogus_dummy')
+        class BogusDummyLicenseFramework:
+            pass
+        ignored(BogusDummyLicenseFramework)
+
+    try:
+        @LicenseFrameworkRegistry.register(name='dummy')
+        class DummyLicenseFramework(LicenseFramework):
+            pass
+
+        dummy_framework = LicenseFrameworkRegistry.LICENSE_FRAMEWORKS.get('dummy')
+        assert issubclass(dummy_framework, LicenseFramework)
+        assert dummy_framework is DummyLicenseFramework
+
+    finally:
+
+        # Clean up the mess we made...
+        for name in ['bogus_dummy', 'dummy']:
+            LicenseFrameworkRegistry.LICENSE_FRAMEWORKS.pop(name, None)
```

Comment:
Better to use a function scope fixture for the registry no?

Reply:
Me? Ha. I never use fixtures except when in a tool that already does, to not fight the style. I hate their scoping rules. Usually I'd reach for a context manager instead because I think their scoping rules are clearer. Or I'd use a class with methods that have natural static scope.  I've rewritten this as a context manager in changes to be uploaded soon.